config/sources/mainline-kernel.conf.sh

```diff
@@ -0,0 +1,63 @@
+function mainline_kernel_decide_version__upstream_release_candidate_number() {
+	[[ -n "${KERNELBRANCH}" ]] && return 0          # if already set, don't touch it; that way other hooks can run in any order
+	if [[ "${KERNEL_MAJOR_MINOR}" == "6.7" ]]; then # @TODO: roll over to 6.8 and v6.8-rc1 when it is released, which should be around Sunday, 2024-01-21 - see https://deb.tandrin.de/phb-crystal-ball.htm
+		declare -g KERNELBRANCH="tag:v6.7-rc7"
+		display_alert "mainline-kernel: upstream release candidate" "Using KERNELBRANCH='${KERNELBRANCH}' for KERNEL_MAJOR_MINOR='${KERNEL_MAJOR_MINOR}'" "info"
+	fi
+}
+
+# Maintained by @viraniac
+function mainline_kernel_decide_version__sunxi_fixed_versions() {
+	[[ "${LINUXFAMILY}" != "sunxi" && "${LINUXFAMILY}" != "sunxi64" ]] && return 0 # only for sunxi/sunxi64
+	case "${KERNEL_MAJOR_MINOR}" in
+		"6.1") declare -g KERNELBRANCH="tag:v6.1.69" ;;
+		"6.6") declare -g KERNELBRANCH="tag:v6.6.8" ;;
+		"6.7") declare -g KERNELBRANCH="tag:v6.7-rc6" ;;
+	esac
+	display_alert "mainline-kernel" "Using fixed version for sunxi/sunxi64 KERNELBRANCH='${KERNELBRANCH}'" "info"
+}
+
+# Maintained by @paolosabatino
+function mainline_kernel_decide_version__rockchip_32bit_fixed_versions() {
+	[[ "${LINUXFAMILY}" != "rockchip" ]] && return 0 # only for rockchip armhf 32-bit
+	case "${KERNEL_MAJOR_MINOR}" in
+		"6.1") declare -g KERNELBRANCH="tag:v6.1.68" ;;
+		"6.6") declare -g KERNELBRANCH="tag:v6.6.7" ;;
+	esac
+	display_alert "mainline-kernel" "Using fixed version for rockchip armhf 32-bit KERNELBRANCH='${KERNELBRANCH}'" "info"
+}
+
+### Later than normal hooks, for emergencies / locking versions for release / etc. Use mainline_kernel_decide_version__600 or higher.
+
+## Example: "6.6.7 was recently released with changes that break our drivers/patches/souls. Let's Lock 6.6 to 6.6.6 until we've time to fix it."
+#function mainline_kernel_decide_version__600_lock_6.6_to_6.6.6() {
+#	[[ "${KERNEL_MAJOR_MINOR}" != "6.6" ]] && return 0 # only for 6.6
+#	declare -g KERNELBRANCH="tag:v6.6.6"
+#	display_alert "mainline-kernel: locked version for 6.6 kernel" "Using fixed version for 6.6 KERNELBRANCH='${KERNELBRANCH}'" "info"
+#}
+
+### Later-than-usual hooks, for changing parameters after the hooks above have run. use mainline_kernel_decide_version__750 or higher.
+
+# https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git is missing 6.7-rc7 tag; use Linus GH repo instead.
+# Attention: this does not support/respect git mirror... hopefully kernel.org catches up soon and we can remove this.
+function mainline_kernel_decide_version__750_use_torvalds_for_6.7-rc7() {
+	if [[ "${KERNELBRANCH}" == 'tag:v6.7-rc7' ]]; then
+		display_alert "Using Linus kernel repo for 6.7-rc7" "${KERNELBRANCH}" "warn"
+		declare -g KERNELSOURCE="https://github.com/torvalds/linux.git"
+		display_alert "mainline-kernel: missing torvalds tag on 6.7-rc7" "Using KERNELSOURCE='${KERNELSOURCE}' for KERNELBRANCH='${KERNELBRANCH}'" "info"
+	fi
+}
+
```

Comment:
Everything above this line is not making much sense to me. All the overrides should be in family configs, No point putting all that into a common file if the same version is not being used. Also I don't like the 6.7-rc7 hack. Its going to be a shortlived code that would be unnecessarily added to the repository. If someone really wish to use 6.7-rc7, then they can just download incremental patch and put it in there kernel patches directory.

Reply:
Please read the whole file including the comments ;-)

Since it is fully composed of hooks (functions), you can move "your" hooks to "your" family file if you so wish (originally, you had it duplicated, since it is the same version for sunxi/sunxi64). Maybe you rather have a common family file for both sunxi's?

The point about v6.7-rc7 is explained in the comment above it, please read it.

Reply:
> The point about v6.7-rc7

Perhaps comment that this is covering some weird exceptions as this is not clear right away.

Do you think this will ever happen again?

Reply:
> Since it is fully composed of hooks (functions), you can move "your" hooks to "your" family file if you so wish (originally, you had it duplicated, since it is the same version for sunxi/sunxi64). Maybe you rather have a common family file for both sunxi's?

I have thought about combining those files a countless times. But then I also remember that some times I have to keep the 32-bit back because it stops booting after kernel bump and I need more time to fix the same. Hence I haven't combined those files into one.

Reply:
Ok, that's a valid allegation -- I never imagined you'd bump arm64 and not armhf. Should I move the version lock back into your 2 families, then? It will be just like before.
@paolosabatino would you also prefer the same for rockchip?

Reply:
> > The point about v6.7-rc7
> 
> Perhaps comment (better) that this is covering some weird exceptions as this is not clear right away.

Added even more comments, hopefully they're "better".

> Do you think this will ever happen again?

It might, or not, I don't know -- I never seen it happen before.


> Should I move the version lock back into your 2 families, then? It will be just like before.

I've moved back both sunxi/sunxi64 and rockchip version locks to their families.

Reply:
_lol_ it seems kernel.org _just_ pushed the missing tag (https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git) There I go removing it...

Reply:
There, commented out and transformed into an example in case it happens again.